### PR TITLE
Explore: Add switch to prettify json logs

### LINKED ISSUE #18522: Explore: Prettify json log
**What would you like to be added**:
Some way to pretty print the searchable json logs in the explore ui. Either through an option in the configuration menu or a checkbox in the explore view itself to enable it globally for all json data.

**Why is this needed**:
Reading big chunks of json data is hard and its easy to get lost in it. If there was a way to prettify the data it would be much easier to review and find the thing that you are searching for in the log line.

Example:

Non prettified data:
```
{"log":"{\"type\":\"response\",\"@timestamp\":\"2019-08-12T20:06:43Z\",\"tags\":[],\"pid\":1,\"method\":\"get\",\"statusCode\":200,\"req\":{\"url\":\"/app/kibana\",\"method\":\"get\",\"headers\":{\"user-agent\":\"curl/7.29.0\",\"host\":\"localhost:5601\",\"accept\":\"*/*\"},\"remoteAddress\":\"127.0.0.1\",\"userAgent\":\"127.0.0.1\"},\"res\":{\"statusCode\":200,\"responseTime\":45,\"contentLength\":9},\"message\":\"GET /app/kibana 200 45ms - 9.0B\"}\n","stream":"stdout","time":"2019-08-12T20:06:43.183719829Z"}
```

Prettified data:
```
{
  "log": "{\"type\":\"response\",\"@timestamp\":\"2019-08-12T20:06:43Z\",\"tags\":[],\"pid\":1,\"method\":\"get\",\"statusCode\":200,\"req\":{\"url\":\"/app/kibana\",\"method\":\"get\",\"headers\":{\"user-agent\":\"curl/7.29.0\",\"host\":\"localhost:5601\",\"accept\":\"*/*\"},\"remoteAddress\":\"127.0.0.1\",\"userAgent\":\"127.0.0.1\"},\"res\":{\"statusCode\":200,\"responseTime\":45,\"contentLength\":9},\"message\":\"GET /app/kibana 200 45ms - 9.0B\"}\n",
  "stream": "stdout",
  "time": "2019-08-12T20:06:43.183719829Z"
}
```

One of the problems with this is that the log field has the log data with escaped ```"``` characters. This makes the inner json structure harder to prettify as one needs to take that in consideration when it is prettified.

Prettified data with unescaped ```"```:
```
{  
  "log":{  
    "type":"response",
    "@timestamp":"2019-08-12T20:06:43Z",
    "tags":[  

    ],
    "pid":1,
    "method":"get",
    "statusCode":200,
    "req":{  
      "url":"/app/kibana",
      "method":"get",
      "headers":{  
        "user-agent":"curl/7.29.0",
        "host":"localhost:5601",
        "accept":"*/*"
      },
      "remoteAddress":"127.0.0.1",
      "userAgent":"127.0.0.1"
    },
    "res":{  
      "statusCode":200,
      "responseTime":45,
      "contentLength":9
    },
    "message":"GET /app/kibana 200 45ms - 9.0B"
  },
  "stream":"stdout",
  "time":"2019-08-12T20:06:43.183719829Z"
}
```

### PR DESCRIPTION
**Pull request by olbo98**
**What this PR does / why we need it**:
This PR adds a switch that toggles whether the json logs will be prettified or not. The message is currently restructured with `JSON.stringify()`.
Switch toggled **off**:
Screenshot 2021-06-29 at 15 59 50

Switch toggled **on**:
Screenshot 2021-06-29 at 16 01 17

Are there any other features that would be nice to have?

**Which issue(s) this PR fixes**:

Fixes #18522

**Special notes for your reviewer**:

### CODE REVIEW

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx [hunk 1]
```diff
@@ -70,11 +80,13 @@ class UnThemedLogRowMessage extends PureComponent<Props> {
       contextIsOpen,
       showContextToggle,
       wrapLogMessage,
+      prettifyLogMessage,
       onToggleContext,
     } = this.props;
 
     const style = getLogRowStyles(theme, row.logLevel);
     const { entry, hasAnsi, raw } = row;
+    const jsonEntry = prettifyLogMessage && isJson(entry) ? JSON.stringify(JSON.parse(entry), undefined, 2) : entry;
```
Comment (Elfo404) [inline packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:89]: not a biggie and i'm not sure about performance implications of parsing multiple times the same string, but we are parsing `entry` twice in the worst case scenario.

can we reduce it to one by doing, for example:
```ts
const jsonEntry = prettifyLogMessage ? prettify(entry) : entry;
```
where `prettify` is something like:
```ts
function prettify(str: string) {
  try {
    return JSON.stringify(JSON.parse(str), undefined, 2)
  } catch {
    return str;
  }
}
```
?

Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:89]: Yeah you're right, i should change that

**Review by ivanahuckova — COMMENTED** [review]
@Elfo404 @olbo98 Just some UX thoughts/brainstorm. If user doesn't have logs in json format, Prettify is not going to do anything. Which possibly could/but doesn't necessarily have to be confusing. What do we think about a pattern, where we would show button/toggle with possibility to prettify only if at least 1 of the logs is in json format. Similarly to `Escape newlines` that only appears if at least 1 log is unescaped.
image

However, possible downside - if user always want to have json prettified logs, there is currently no pattern on how to persist this state.  Also I can see that in logs panel someone would want this feature and would like to see json logs always prettified. What are your thoughts?

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx (hunk 1 above)
Comment (hugohaggmark) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:89]: There is already such a function :) https://github.com/grafana/grafana/blob/abb1c107767e8011b44078f1212ce59cf718374c/public/app/core/utils/explore.ts#L195

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx [hunk 2]
```diff
@@ -52,6 +53,15 @@ const getStyles = stylesFactory((theme: GrafanaTheme) => {
...
 
+function isJson(str: string) {
+  try {
+    JSON.parse(str);
+  } catch (e) {
+    return false;
+  }
+  return true;
+}
+
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:64]: Consider moving this function next to https://github.com/grafana/grafana/blob/abb1c107767e8011b44078f1212ce59cf718374c/public/app/core/utils/explore.ts#L183

or maybe use that function?

**Review by hugohaggmark — COMMENTED** [review]
Haven't tested this but did a shallow code review and left some minor comments.

**Comment by olbo98** [comment]
> @Elfo404 @olbo98 Just some UX thoughts/brainstorm. If user doesn't have logs in json format, Prettify is not going to do anything. Which possibly could/but doesn't necessarily have to be confusing. What do we think about a pattern, where we would show button/toggle with possibility to prettify only if at least 1 of the logs is in json format. Similarly to `Escape newlines` that only appears if at least 1 log is unescaped.
> image
> 
> However, possible downside - if user always want to have json prettified logs, there is currently no pattern on how to persist this state. Also I can see that in logs panel someone would want this feature and would like to see json logs always prettified. What are your thoughts?

@ivanahuckova I think that's a great idea! Sometimes i think i's a little bit confusing to only display buttons in certain cases. What do you think  about making the switch grey or "disabled" when there aren't any logs in JSON format? If users always want prettified logs maybe we could add an option for that in the datasource settings?

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx (hunk 2 above)
Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx]: Didn't know there was such a function, thank you for the tip! I'll use that one instead

**Comment by olbo98** [comment]
@grafana/observability-squad does anyone know why the build is failing. I tried to rebase on main but it didn't work. I've checked out the files that the build is failing on but everything seems okay and there are no errors in the code. I have no idea what to do 🥲

**Comment by olbo98** [comment]
@hugohaggmark there seems to be a bug with `safeStringifyValye` in my case. When using it on my logs it just adds quotation marks around the logs. I think this happens because of this line 
```
return JSON.stringify(value, null, space);
``` 
We're just passing in the value without parsing it. My suggestion is to change it to 
```
return JSON.stringify(JSON.parse(value), null, space);
```
This seems to work at least on my instance. Is this a reasonable solution or will it break other parts of grafana?

**Comment by ivanahuckova** [comment]
I think you rebased incorrectly because there are quite a lot of changes in many files (~28). I usually just do `git merge origin main` and if there are any merge conflict, resolve those. If there are no merge conflicts, after merging main, that merge commit shows no changes

**Comment by hugohaggmark** [comment]
> @hugohaggmark there seems to be a bug with `safeStringifyValye` in my case. When using it on my logs it just adds quotation marks around the logs. I think this happens because of this line
> 
> ```
> return JSON.stringify(value, null, space);
> ```
> 
> We're just passing in the value without parsing it. My suggestion is to change it to
> 
> ```
> return JSON.stringify(JSON.parse(value), null, space);
> ```
> 
> This seems to work at least on my instance. Is this a reasonable solution or will it break other parts of grafana?

I don't think we should change that behavior, it's used in various places. Maybe you should call `safeParseJson` before calling `safeStringifyValue`? Or decide depending on the type of value? But do that outside of that function.

**Comment by olbo98** [comment]
@ivanahuckova oh okay i just did `git rebase main` which should've worked. Is there any way to revert it?

**Comment by ivanahuckova** [comment]
Maybe https://stackoverflow.com/questions/13514182/git-how-to-revert-a-rebase-that-has-been-pushed-to-the-remote-branch-not-orig/13514702

File: docs/sources/alerting/unified-alerting/contact-points.md [hunk 3]
```diff
@@ -54,20 +54,20 @@ VictorOps | `victorops`
...
+## Managing contact points for an external alertmanager
 
-Grafana alerting UI supports managing external Alertmanager configuration. Once you add an [Alertmanager data source]({{< relref "../../datasources/alertmanager.md" >}}), a dropdown displays at the top of the page where you can select either `Grafana` or an external Alertmanager as your data source. 
+Grafana alerting UI supports managing external alertmanager configuration. Add an [alertmanager datasource]({{< relref "../../datasources/alertmanager.md" >}}) and a dropdown will appear at the top of the page, allowing you to select either `Grafana` or an external alertmanager. 
 
-{{< figure max-width="40%" src="/static/img/docs/alerting/unified/contact-points-select-am-8-0.gif" caption="Select Alertmanager" >}}
+{{< figure max-width="40%" src="/static/img/docs/alerting/unified/contact-points-select-am-8-0.gif" caption="Select alertmanager" >}}
 
-### Edit Alertmanager global config
+### Editing alertmanager global config
```
Comment (achatterjee-grafana) [inline docs/sources/alerting/unified-alerting/contact-points.md:63]: Please keep "Edit".

File: docs/sources/alerting/unified-alerting/contact-points.md [hunk 4]
```diff
@@ -54,20 +54,20 @@ VictorOps | `victorops`
 [Webhook](#webhook) | `webhook`
 [Zenduty](#zenduty) | `webhook` 
 
-## Manage contact points for an external Alertmanager
+## Managing contact points for an external alertmanager
```
Comment (achatterjee-grafana) [inline docs/sources/alerting/unified-alerting/contact-points.md:57]: Please keep "Manage". 
Also, I thought "Alertmanager" is a feature and we need to keep it with uppercase A?

File: docs/sources/alerting/unified-alerting/contact-points.md [hunk 5]
```diff
@@ -54,20 +54,20 @@ VictorOps | `victorops`
...
 
-### Edit Alertmanager global config
+### Editing alertmanager global config
 
 To edit global configuration options for an alertmanager, like SMTP server that is used by default for all email contact types:
 
 
 1. In the Grafana side bar, hover your cursor over the **Alerting** (bell) icon and then click **Contact points**.
-1. In the dropdown at the top of the page, select an Alertmanager data source.
+1. In the dropdown at the top of the page, select a alertmanager datasource.
```
Comment (achatterjee-grafana) [inline docs/sources/alerting/unified-alerting/contact-points.md:69]: Why a alertmanager? We should use the article "an" here.

**Review by achatterjee-grafana — COMMENTED** [review]
Why was "Template data" and "Message templating" topics removed? Also added a few comments.

**Comment by olbo98** [comment]
@achatterjee-grafana these changes were added because something went wrong when i tried to rebase my branch. I'll have to close this PR and create a new one 😅

**Comment by olbo98** [comment]
Closing this PR because something went wrong with rebasing. Will create a new PR

**Pull request closed without merging**